skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +255,350 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    unique_nodes = np.logical_xor(u_node_counts, v_node_counts)
+    observed_nodes = np.logical_or(u_node_counts, v_node_counts)
+    unique_branch_length = (branch_lengths * unique_nodes).sum()
+    observed_branch_length = (branch_lengths * observed_nodes).sum()
+    if observed_branch_length == 0.0:
+        # handle special case to avoid division by zero
         return 0.0
-    weighted_unifrac = 0
-    D = 0
-    for o in uv_obs_nodes:
-        # handle the case of o.length is None
-        b = o.length or 0
-        u_branch_weight = _sample_branch_weight(u_obs_nodes[o], u_total_count)
-        v_branch_weight = _sample_branch_weight(v_obs_nodes[o], v_total_count)
-        branch_weight = abs(u_branch_weight - v_branch_weight)
-        weighted_unifrac += b * branch_weight
-        if normalized and o.is_tip():
-            d = o.accumulate_to_ancestor(tree.root())
-            normed_weight = u_branch_weight + v_branch_weight
-            D += (d * normed_weight)
-    if normalized:
-        return weighted_unifrac / D
+    return unique_branch_length / observed_branch_length
+
+
+def _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                      v_total_count, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+        The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+        respectively. This could be computed internally, but since this is a
+        private method and the calling function has already generated these
+        values, this saves an iteration over each of these vectors.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        Weighted UniFrac distance between samples.
+    np.array of float
+        Proportional abundance of each node in tree in sample `u`
+    np.array of float
+        Proportional abundance of each node in tree in sample `v`
+
+    """
+    if u_total_count > 0:
+        # convert to relative abundances if there are any counts
+        u_node_proportions = u_node_counts / u_total_count
     else:
-        return weighted_unifrac
+        # otherwise, we'll just do the computation with u_node_counts, which
+        # is necessarily all zeros
+        u_node_proportions = u_node_counts
+
+    if v_total_count > 0:
+        v_node_proportions = v_node_counts / v_total_count
+    else:
+        v_node_proportions = v_node_counts
+
+    wu = (branch_lengths * abs(u_node_proportions - v_node_proportions)).sum()
```

Comment:
good catch

On Thu, Nov 19, 2015 at 12:28 PM, Jai Ram Rideout <notifications@github.com>
wrote:

> In skbio/diversity/beta/_unifrac.py
> <https://github.com/biocore/scikit-bio/pull/1040#discussion_r45395834>:
>
> >      else:
> > -        return weighted_unifrac
> > +        # otherwise, we'll just do the computation with u_node_counts, which
> > +        # is necessarily all zeros
> > +        u_node_proportions = u_node_counts
> > +
> > +    if v_total_count > 0:
> > +        v_node_proportions = v_node_counts / v_total_count
> > +    else:
> > +        v_node_proportions = v_node_counts
> > +
> > +    wu = (branch_lengths * abs(u_node_proportions - v_node_proportions)).sum()
>
> probably faster to use np.absolute instead of abs
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1040/files#r45395834>.
>